components/application-deployer/org.wso2.carbon.event.application.deployer/src/main/java/org/wso2/carbon/event/application/deployer/EventProcessingAppDeployer.java

```diff
@@ -119,42 +126,62 @@ private void addArtifact(Artifact artifact) {
+            undeployTypeSpecifiedArtifacts(eventPublishers, axisConfig, EventProcessingAppDeployerConstants.CEP_EVENT_PUBLISHER_DIR,
+                    EventProcessingAppDeployerConstants.FILE_TYPE_XML);
+            undeployTypeSpecifiedArtifacts(executionPlans, axisConfig, EventProcessingAppDeployerConstants.CEP_EXECUTION_PLAN_DIR,
+                    EventProcessingAppDeployerConstants.FILE_TYPE_SIDDHIQL);
+        } finally {
+            executionPlans.clear();
+            eventStreams.clear();
+            eventReceivers.clear();
+            eventPublishers.clear();
+        }
+    }
 
-            List<CappFile> files = artifact.getFiles();
-            if (files.size() != 1) {
-                log.error("Artifact must have only a single file. But " +
-                        files.size() + " files found.");
-                continue;
+    private void undeployTypeSpecifiedArtifacts(List<Artifact> artifacts, AxisConfiguration axisConfig, String directory, String fileType) throws DeploymentException {
+        for(Artifact artifact: artifacts) {
+            EventProcessingDeployer deployer;
+            deployer = (EventProcessingDeployer)AppDeployerUtils.getArtifactDeployer(axisConfig,
+                    directory, fileType);
+            if(deployer!=null) {
+                undeploy(deployer, artifact);
             }
+        }
+    }
 
-            if (deployer != null && AppDeployerConstants.DEPLOYMENT_STATUS_DEPLOYED.
-                    equals(artifact.getDeploymentStatus())) {
-                String fileName = artifact.getFiles().get(0).getName();
-                String artifactPath = artifact.getExtractedPath() + File.separator + fileName;
-                try {
-                    deployer.processUndeployment(artifactPath);
-                    artifact.setDeploymentStatus(AppDeployerConstants.DEPLOYMENT_STATUS_PENDING);
-                    File artifactFile = new File(artifactPath);
-                    if (artifactFile.exists() && !artifactFile.delete()) {
-                        log.warn("Couldn't delete App artifact file : " + artifactPath);
-                    }
-                } catch (Exception e) {
-                    artifact.setDeploymentStatus(AppDeployerConstants.DEPLOYMENT_STATUS_FAILED);
-                    log.error("Error occured while trying to undeploy : " + artifact.getName() + " due to " + e.getMessage(), e);
-                }
-            }
+    private void undeploy(EventProcessingDeployer deployer, Artifact artifact) throws DeploymentException {
+        List<CappFile> files = artifact.getFiles();
+        String fileName = artifact.getFiles().get(0).getName();
+        String artifactPath = artifact.getExtractedPath() + File.separator + fileName;
+        try {
+            deployer.processUndeployment(new DeploymentFileData(new File(artifactPath), deployer).getAbsolutePath());
+            artifact.setDeploymentStatus(AppDeployerConstants.DEPLOYMENT_STATUS_DEPLOYED);
```

Comment:
yeah it should be DEPLOYMENT_STATUS_PENDING, not failed. I don't know why that has to be pending either, but, that is how it is written in the other app deployers.

It really undeploys only the things previously deployed. The undeployment also follows the same artifact order of deployment. So, once it tries to undeploy the things previously not deployed, it will throw an error and that error will be caught in line 79 catch block. Here also we have an error that has to be ignored after catching since it is an artificially generated error.

Thanks for the comments. I'll change them and have another pull request.